x-pack/plugin/identity-provider/src/test/java/org/elasticsearch/xpack/idp/privileges/UserPrivilegeResolverTests.java

```diff
@@ -0,0 +1,231 @@
+        assertThat(privileges.hasAccess, equalTo(true));
+        assertThat(privileges.groups, containsInAnyOrder("viewer"));
+    }
+
+    public void testResolveSsoWithMultipleGroup() throws Exception {
+        final String username = randomAlphaOfLengthBetween(4, 12);
+        final String app = randomAlphaOfLengthBetween(3, 8);
+        final String resource = "cluster:" + MessageDigests.toHexString(randomByteArrayOfLength(16));
+        final String loginAction = "action:idp:sso";
+        final String viewerAction = "role:cluster:view";
+        final String adminAction = "role:cluster:admin";
+        final String operatorAction = "role:cluster:operator";
+        final String monitorAction = "role:cluster:monitor";
+
+        setupUser(username);
+        setupHasPrivileges(username, app,
+            access(resource, loginAction, true),
+            access(resource, viewerAction, false),
+            access(resource, adminAction, false),
+            access(resource, operatorAction, true),
+            access(resource, monitorAction, true)
+        );
+
+        final PlainActionFuture<UserPrivilegeResolver.UserPrivileges> future = new PlainActionFuture<>();
+        resolver.resolve(service(app, resource, loginAction,
+            Map.of("viewer", viewerAction, "admin", adminAction, "operator", operatorAction, "monitor", monitorAction)),
+            future);
+        final UserPrivilegeResolver.UserPrivileges privileges = future.get();
+        assertThat(privileges.principal, equalTo(username));
+        assertThat(privileges.hasAccess, equalTo(true));
+        assertThat(privileges.groups, containsInAnyOrder("operator", "monitor"));
+    }
+
+    public void testResolveGroupAccessButNoSsoPrivilege() throws Exception {
+        final String username = randomAlphaOfLengthBetween(4, 12);
+        final String app = randomAlphaOfLengthBetween(3, 8);
+        final String resource = "cluster:" + MessageDigests.toHexString(randomByteArrayOfLength(16));
+        final String loginAction = "action:idp:sso";
+        final String viewerAction = "role:cluster:view";
+        final String adminAction = "role:cluster:admin";
+
+        setupUser(username);
+        setupHasPrivileges(username, app,
+            access(resource, loginAction, false), access(resource, viewerAction, true), access(resource, adminAction, false));
+
+        final PlainActionFuture<UserPrivilegeResolver.UserPrivileges> future = new PlainActionFuture<>();
+        resolver.resolve(service(app, resource, loginAction, Map.of("viewer", viewerAction, "admin", adminAction)), future);
+        final UserPrivilegeResolver.UserPrivileges privileges = future.get();
+        assertThat(privileges.principal, equalTo(username));
+        assertThat(privileges.hasAccess, equalTo(false));
+        assertThat(privileges.groups, containsInAnyOrder("viewer"));
+    }
+
+    public void testResolveSsoWithNoGroups() throws Exception {
```

Comment:
Duplicate of `testResolveSsoWithNoGroups`

Reply:
This is `testResolveSsoWithNoGroups`. Do you mean duplicate of `testResolveSsoWithNoGroupsDefined`.
In which case the 2 tests do different things (but this one was supposed to get renamed, and I missed it).

`testResolveSsoWithNoGroupsDefined` tests when an SP does not declare any group assignment privileges.
`testResolveSso**WhenUserHas**NoGroups` (as it should be named) tests when an SP declares group privileges, but the user has none of them.

Reply:
Oh, wait there is a duplicate. 
We already have `testResolveSsoWithNoGroupAccess` and `testResolveSsoWithNoGroupsDefined` this was left over from when I split it out into those 2 separate cases.